File: app/src/test/java/info/nightscout/androidaps/plugins/general/automation/actions/ActionSendSMSTest.java
package info.nightscout.androidaps.plugins.general.automation.actions;

import android.telephony.SmsManager;

import com.google.common.base.Optional;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import info.AAPSMocker;
import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.plugins.general.automation.elements.InputString;
import info.nightscout.androidaps.queue.Callback;
import info.nightscout.androidaps.utils.SP;

import static org.powermock.api.mockito.PowerMockito.mock;
import static org.powermock.api.mockito.PowerMockito.mockStatic;

@RunWith(PowerMockRunner.class)
@PrepareForTest({MainApp.class, SP.class, SmsManager.class})
public class ActionSendSMSTest {
    private ActionSendSMS actionSendSMS;

    @Test
    public void friendlyNameTest() {
        Assert.assertEquals(R.string.sendsmsactiondescription, actionSendSMS.friendlyName());
    }

    @Test
    public void shortDescriptionTest() {
        actionSendSMS = new ActionSendSMS();
        Assert.assertEquals(null, actionSendSMS.shortDescription()); // not mocked
    }

    @Test
    public void iconTest() {
        Assert.assertEquals(Optional.of(R.drawable.ic_notifications), actionSendSMS.icon());
    }

    @Test
    public void doActionTest() {
        actionSendSMS.text = new InputString().setValue("Asd");
        actionSendSMS.doAction(new Callback() {
            @Override
            public void run() {
                Assert.assertTrue(result.success);
            }
        });
    }

    @Test
    public void hasDialogTest() {
        Assert.assertTrue(actionSendSMS.hasDialog());
    }

    @Test
    public void toJSONTest() {
        actionSendSMS = new ActionSendSMS();
        actionSendSMS.text = new InputString().setValue("Asd");
        Assert.assertEquals("{\"data\":{\"text\":\"Asd\"},\"type\":\"info.nightscout.androidaps.plugins.general.automation.actions.ActionSendSMS\"}", actionSendSMS.toJSON());
    }

    @Test
    public void fromJSONTest() {
        actionSendSMS = new ActionSendSMS();
        actionSendSMS.fromJSON("{\"text\":\"Asd\"}");
        Assert.assertEquals("Asd", actionSendSMS.text.getValue());
    }

    @Before
    public void prepareTest() {
        AAPSMocker.mockMainApp();
        AAPSMocker.mockSP();
        mockStatic(SmsManager.class);
        SmsManager smsManager = mock(SmsManager.class);
        PowerMockito.when(SmsManager.getDefault()).thenReturn(smsManager);
        PowerMockito.when(SP.getString(R.string.key_smscommunicator_allowednumbers, "")).thenReturn("1234;5678");
        actionSendSMS = new ActionSendSMS();
    }
}

File: app/src/test/java/info/nightscout/androidaps/plugins/general/automation/actions/ActionSendSMSTest.kt
package info.nightscout.androidaps.plugins.general.automation.actions

import android.widget.LinearLayout
import info.nightscout.androidaps.R
import info.nightscout.androidaps.plugins.general.automation.elements.InputString
import info.nightscout.androidaps.queue.Callback
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyString
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mockito
import org.mockito.Mockito.`when`
import org.powermock.modules.junit4.PowerMockRunner

@RunWith(PowerMockRunner::class)
class ActionSendSMSTest : ActionsTestBase() {

    private lateinit var sut: ActionSendSMS

    @Before
    fun setup() {

        `when`(resourceHelper.gs(eq(R.string.sendsmsactionlabel), anyString())).thenReturn("Send SMS: %s")
        `when`(resourceHelper.gs(R.string.sendsmsactiondescription)).thenReturn("Send SMS to all numbers")

        sut = ActionSendSMS(injector)
    }

    @Test fun friendlyNameTest() {
        Assert.assertEquals(R.string.sendsmsactiondescription, sut.friendlyName())
    }

    @Test fun shortDescriptionTest() {
        Assert.assertEquals("Send SMS: %s", sut.shortDescription())
    }

    @Test fun iconTest() {
        Assert.assertEquals(R.drawable.ic_notifications, sut.icon())
    }

    @Test fun doActionTest() {
        `when`(smsCommunicatorPlugin.sendNotificationToAllNumbers(anyString())).thenReturn(true)
        sut.text = InputString(injector, "Asd")
        sut.doAction(object : Callback() {
            override fun run() {
                Assert.assertTrue(result.success)
            }
        })
    }

    @Test fun hasDialogTest() {
        Assert.assertTrue(sut.hasDialog())
    }

    @Test fun toJSONTest() {
        sut.text = InputString(injector, "Asd")
        Assert.assertEquals("{\"data\":{\"text\":\"Asd\"},\"type\":\"info.nightscout.androidaps.plugins.general.automation.actions.ActionSendSMS\"}", sut.toJSON())
    }

    @Test fun fromJSONTest() {
        sut.fromJSON("{\"text\":\"Asd\"}")
        Assert.assertEquals("Asd", sut.text.value)
    }
}